component/org.wso2.carbon.identity.dpop/src/main/java/org/wso2/carbon/identity/dpop/handler/DPoPAuthenticationHandler.java

```diff
@@ -0,0 +1,343 @@
+
+                    getAuthenticationResult(authenticationResult, responseDTO, authorizationHeader,
+                            authenticationRequest, messageContext, accessToken, consumerKey);
+
+                    setAuthenticationContext(responseDTO, authenticationContext, consumerKey);
+
+                    String serviceProvider;
+                    try {
+                        serviceProvider =
+                                OAuth2Util.getServiceProvider(consumerKey).getApplicationName();
+                    } catch (IdentityOAuth2Exception e) {
+                        String error = String.format("Error occurred while getting the Service Provider" +
+                                " by Consumer key: %s.", consumerKey);
+                        log.error(error, e);
+                        throw new AuthenticationFailException(error);
+                    }
+
+                    String serviceProviderTenantDomain;
+                    try {
+                        serviceProviderTenantDomain =
+                                OAuth2Util.getTenantDomainOfOauthApp(consumerKey);
+                    } catch (InvalidOAuthClientException | IdentityOAuth2Exception e) {
+
+                        String error = String.format("Error occurred while getting the OAuth App" +
+                                " tenantDomain by Consumer key: %s.", consumerKey);
+                        log.error(error, e);
+                        throw new AuthenticationFailException(error);
+                    }
+
+                    authenticationContext.addParameter(DPoPConstants.SERVICE_PROVIDER, serviceProvider);
+                    authenticationContext.addParameter(DPoPConstants.SERVICE_PROVIDER_TENANT_DOMAIN,
+                            serviceProviderTenantDomain);
+
+                    MDC.put(DPoPConstants.SERVICE_PROVIDER, serviceProvider);
+                    // Set OAuth service provider details to be consumed by the provisioning framework.
+                    setProvisioningServiceProviderThreadLocal(clientApplicationDTO.getConsumerKey(),
+                            serviceProviderTenantDomain);
+
+                } else {
+                    String errorMessage = String.format("Error occurred while trying to authenticate." +
+                            "The %s header value is not defined correctly.", DPoPConstants.OAUTH_DPOP_HEADER);
+                    log.error(errorMessage);
+                    throw new AuthenticationFailException(errorMessage);
+                }
+            }
+        }
+        return authenticationResult;
+    }
+
+    @Override
+    public int getPriority(MessageContext messageContext) {
+
+        return getPriority(messageContext, 24);
+    }
+
+    @Override
+    public boolean canHandle(MessageContext messageContext) {
+
+        return AuthConfigurationUtil.isAuthHeaderMatch(messageContext, DPoPConstants.OAUTH_DPOP_HEADER) ||
```

Comment:
Do we need to handle OAuth Header here. If that header comes, that can be handled by the default OAuth2AccessTokenHandler. DPoP specific authentication logic should come here

Reply:
If we remove the OAuth Header DPoP tokens can be used in the Authorization header with the Bearer prefix when invoking protected APIs.

Reply:
Fixed